### PR TITLE
Playwright Extended Support for Network Log ,Alert message and interp…

### PR DESCRIPTION
**Pull request by prashelke**
…olation added in Analyzer

Thank you for your contribution.
Before submitting this PR, please make sure:

- [x] PR description and commit message should describe the changes done in this PR
- [x] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [x] Latest Code from master or specific release branch is merged to your branch
- [x] No unwanted\commented\junk code is included
- [x] No new warning upon build solution
- [x] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [x] Sanity Tests are successfully executed for New and Existing Functionality
- [x] Verify that changes are compatible with all relevant browsers and platforms.
- [x] After creating pull request there should not be any conflicts
- [ ] Resolve all Codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Update the Help Library document to match any feature changes

## Release Notes

- **New Features**
  - Enhanced support for handling message boxes and network logging in the browser automation framework.
  - Introduced methods for monitoring network logs and interacting with dialog boxes, including accepting and dismissing message boxes.
  - Added a new `BrowserHelper` class for URL monitoring and network log file creation.

- **Bug Fixes**
  - Improved error handling for network log operations and message box interactions, ensuring better reliability and clearer feedback.
  - Enhanced error messaging for unsupported actions and conditions within the browser driver.

### CODE REVIEW

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightDriver.cs [hunk 1]
```diff
@@ -1810,6 +1810,13 @@ public string GetViewport()
             ActBrowserElement.eControlAction.SwitchToParentFrame,
             ActBrowserElement.eControlAction.SwitchWindow,
             ActBrowserElement.eControlAction.SwitchToDefaultWindow,
+            ActBrowserElement.eControlAction.AcceptMessageBox,
+            ActBrowserElement.eControlAction.DismissMessageBox,
+            ActBrowserElement.eControlAction.GetMessageBoxText,
+            ActBrowserElement.eControlAction.SetAlertBoxText,
+            ActBrowserElement.eControlAction.StartMonitoringNetworkLog,
+            ActBrowserElement.eControlAction.GetNetworkLog,
+            ActBrowserElement.eControlAction.StopMonitoringNetworkLog,
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightDriver.cs:1819]: will handled in next time

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs [hunk 2]
```diff
@@ -605,6 +621,237 @@ public static bool IsFrameLocatorSupported(eLocateBy locateBy)
...
+                    Act.AddArtifactToAction(Path.GetFileName(responsePath), act, responsePath);
+                });
+            }
+            catch (Exception ex)
+            {
+                Reporter.ToLog(eLogLevel.ERROR, $"Method - {MethodBase.GetCurrentMethod().Name}, Error - {ex.Message}", ex);
+            }
+    
+        }
+
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs:735]: updated

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs [hunk 3]
```diff
@@ -605,6 +621,237 @@ public static bool IsFrameLocatorSupported(eLocateBy locateBy)
...
+                            networkResponseLogList.Add(new Tuple<string, object>($"ResponseUrl:{response.Url}", JsonConvert.SerializeObject(response, Formatting.Indented,
+                                                                                                                                    new JsonSerializerSettings
+                                                                                                                                    {
+                                                                                                                                        ReferenceLoopHandling = ReferenceLoopHandling.Ignore
+                                                                                                                                    })
+                            ));
+                        }
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs:801]: not need to Response.BodyAsync()

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs [hunk 4]
```diff
@@ -605,6 +621,237 @@ public static bool IsFrameLocatorSupported(eLocateBy locateBy)
...
+            try
+            {
+                await dialogs.AcceptAsync(promptText:MessageBoxText);
+            }
+            catch (Exception ex)
+            {
+                Reporter.ToLog(eLogLevel.ERROR, "Error While Accept Message", ex);
+            }
+
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs [hunk 5]
```diff
@@ -605,6 +621,237 @@ public static bool IsFrameLocatorSupported(eLocateBy locateBy)
...
+            {
+                await Task.Run(() => {
+                    dialogs = e;
+                });
+            }
+            catch (Exception ex)
+            {
+                Reporter.ToLog(eLogLevel.ERROR, $"Method - {MethodBase.GetCurrentMethod().Name}, Error - {ex.Message}", ex);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs:824]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs [hunk 6]
```diff
@@ -538,5 +567,102 @@ private async Task HandleSwitchToDefaultWindowOperationAsync()
...
+        {
+            try
+            {
+                ((PlaywrightBrowserTab)_browser!.CurrentWindow.CurrentTab).StopCaptureNetworkLog(_act);
+            }
+            catch (Exception e)
+            {
+                _act.Error = "Error when Accepting MessageBox.";
+                Reporter.ToLog(eLogLevel.ERROR, "Error when Accepting MessageBox - " + e.Message);
+                return;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs]: added

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs [hunk 7]
```diff
@@ -538,5 +567,102 @@ private async Task HandleSwitchToDefaultWindowOperationAsync()
...
+        {
+            try
+            {
+                await ((PlaywrightBrowserTab)_browser!.CurrentWindow.CurrentTab).GetCaptureNetworkLog(_act);
+            }
+            catch (Exception e)
+            {
+                _act.Error = "Error when Accepting MessageBox.";
+                Reporter.ToLog(eLogLevel.ERROR, "Error when Accepting MessageBox - " + e.Message);
+                return;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs]: added

[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs]: added

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs [hunk 8]
```diff
@@ -538,5 +567,102 @@ private async Task HandleSwitchToDefaultWindowOperationAsync()
...
+        {
+            try
+            {
+               await ((PlaywrightBrowserTab)_browser!.CurrentWindow.CurrentTab).StartCaptureNetworkLog(_act);
+            }
+            catch (Exception e)
+            {
+                _act.Error = "Error when Accepting MessageBox.";
+                Reporter.ToLog(eLogLevel.ERROR, "Error when Accepting MessageBox - " + e.Message);
+                return;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs]: added

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/BrowserHelper.cs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+using System.Threading.Tasks;
+
+namespace Amdocs.Ginger.CoreNET.Drivers.CoreDrivers.Web
+{
+    internal class BrowserHelper
+    {
+        ActBrowserElement _act;
+        public BrowserHelper() {
+
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/BrowserHelper.cs]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/BrowserHelper.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Amdocs.Ginger.CoreNET.Drivers.CoreDrivers.Web
+{
+    internal class BrowserHelper
+    {
+        ActBrowserElement _act;
+        public BrowserHelper() {
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/BrowserHelper.cs]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/BrowserHelper.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+            _act = act;
+            return _act.GetOrCreateInputParam(nameof(ActBrowserElement.eMonitorUrl)).Value == nameof(ActBrowserElement.eMonitorUrl.AllUrl);
+        }
+
+        public bool IsToMonitorOnlySelectedUrls(ActBrowserElement act,string requestUrl)
+        {
+            return _act.GetOrCreateInputParam(nameof(ActBrowserElement.eMonitorUrl)).Value == nameof(ActBrowserElement.eMonitorUrl.SelectedUrl)
+                && _act.UpdateOperationInputValues != null
+                && _act.UpdateOperationInputValues.Any(x => !string.IsNullOrEmpty(x.ValueForDriver) && requestUrl.ToLower().Contains(x.ValueForDriver.ToLower()));
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/BrowserHelper.cs]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/BrowserHelper.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+                using (Stream fileStream = System.IO.File.Create(FullFilePath))
+                {
+                    fileStream.Close();
+                }
+                System.IO.File.WriteAllText(FullFilePath, FileContent);
+
+            }
+            return FullFilePath;
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/BrowserHelper.cs]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs [hunk 13]
```diff
@@ -538,5 +568,132 @@ private async Task HandleSwitchToDefaultWindowOperationAsync()
...
+        private async Task HandleAcceptMessageBoxOperationAsync()
+        {
+            try
+            {
+                await ((PlaywrightBrowserTab)_browser!.CurrentWindow.CurrentTab).AcceptMessageBox();
+            }
+            catch (Exception ex)
+            {
+                _act.Error = $"Error when {MethodBase.GetCurrentMethod().Name}.";
+                Reporter.ToLog(eLogLevel.ERROR, $"Method - {MethodBase.GetCurrentMethod().Name}, Error - {ex.Message}", ex);
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs:585]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs [hunk 14]
```diff
@@ -538,5 +568,132 @@ private async Task HandleSwitchToDefaultWindowOperationAsync()
...
+        /// <summary>
+        /// This asynchronous method accepts a message box (such as a dialog box) in the current browser tab. 
+        /// If an error occurs, it logs the error and updates the Error property.
+        /// </summary>
+        /// <returns></returns>
+        private async Task HandleAcceptMessageBoxOperationAsync()
+        {
+            try
+            {
+                await ((PlaywrightBrowserTab)_browser!.CurrentWindow.CurrentTab).AcceptMessageBox();
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs:580]: not need

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs [hunk 15]
```diff
@@ -538,5 +568,132 @@ private async Task HandleSwitchToDefaultWindowOperationAsync()
...
+        private async Task HandleStartMonitoringNetworkLogOperationAsync()
+        {
+            try
+            {
+               await ((PlaywrightBrowserTab)_browser!.CurrentWindow.CurrentTab).StartCaptureNetworkLog(_act);
+            }
+            catch (Exception ex)
+            {
+                _act.Error = $"Error when {MethodBase.GetCurrentMethod().Name}.";
+                Reporter.ToLog(eLogLevel.ERROR, $"Method - {MethodBase.GetCurrentMethod().Name}, Error - {ex.Message}", ex);
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs:658]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs [hunk 16]
```diff
@@ -538,5 +568,132 @@ private async Task HandleSwitchToDefaultWindowOperationAsync()
...
+        /// <summary>
+        /// This asynchronous method stops monitoring and capturing network logs in the current browser tab. 
+        /// If an error occurs, it logs the error and updates the Error property.
+        /// </summary>
+        /// <returns></returns>
+        private async Task HandleStopMonitoringNetworkLogOperationAsync()
+        {
+            try
+            {
+                ((PlaywrightBrowserTab)_browser!.CurrentWindow.CurrentTab).StopCaptureNetworkLog(_act);
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs [hunk 17]
```diff
@@ -145,6 +147,34 @@ internal async Task HandleAsync()
...
+                    case ActBrowserElement.eControlAction.DismissMessageBox:
+                        await HandleDismissMessageBoxOperationAsync();
+                        break;
+                    case ActBrowserElement.eControlAction.GetMessageBoxText:
+                        string AlertBoxText = HandleGetMessageBoxTextOperation();
+                        _act.AddOrUpdateReturnParamActual("Actual", AlertBoxText);
+                        if (_act.GetReturnParam("Actual") == null)
+                        {
+                            _act.AddOrUpdateReturnParamActual("Actual", AlertBoxText);
+                        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActBrowserElementHandler.cs:162]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs [hunk 18]
```diff
@@ -605,6 +621,274 @@ public static bool IsFrameLocatorSupported(eLocateBy locateBy)
...
+            {
+                return dialogs.Message;
+            }
+            catch (Exception ex)
+            {
+                Reporter.ToLog(eLogLevel.ERROR, $"Method - {MethodBase.GetCurrentMethod().Name}, Error - {ex.Message}", ex);
+                return "Error While Get Message Box Text";
+            }
+
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs [hunk 19]
```diff
@@ -605,6 +621,274 @@ public static bool IsFrameLocatorSupported(eLocateBy locateBy)
...
+                    _playwrightPage.Response += OnNetworkResponseReceived;
+                    isNetworkLogMonitoringStarted = true;
+                });
+            }
+            catch(Exception ex)
+            {
+                Reporter.ToLog(eLogLevel.ERROR, $"Method - {MethodBase.GetCurrentMethod().Name}, Error - {ex.Message}", ex);
+            }
+            
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs [hunk 20]
```diff
@@ -605,6 +621,305 @@ public static bool IsFrameLocatorSupported(eLocateBy locateBy)
...
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs [hunk 21]
```diff
@@ -63,8 +70,15 @@ internal sealed class PlaywrightBrowserTab : IBrowserTab
         private readonly LinkedList<string> _consoleMessages = [];
         private IFrame _currentFrame;
         private bool _isClosed = false;
+        List<Tuple<string, object>> networkResponseLogList;
+        List<Tuple<string, object>> networkRequestLogList;
+        ObservableList<IResponse> NetworkResponseList;
+        ActBrowserElement _act;
+        public bool isNetworkLogMonitoringStarted = false;
+        IDialog dialogs;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs:79]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs [hunk 22]
```diff
@@ -605,6 +622,302 @@ public static bool IsFrameLocatorSupported(eLocateBy locateBy)
...
+                        act.ExInfo = $"Action is skipped,{nameof(ActBrowserElement.eControlAction.StartMonitoringNetworkLog)} Action is not started";
+                        act.Status = Amdocs.Ginger.CoreNET.Execution.eRunStatus.Skipped;
+                    }
+                });
+            }
+            catch (Exception ex)
+            {
+                Reporter.ToLog(eLogLevel.ERROR, $"Method - {MethodBase.GetCurrentMethod().Name}, Error - {ex.Message}", ex);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs [hunk 23]
```diff
@@ -10930,49 +10931,64 @@ private void blockOrUnblockUrls()
...
-            foreach (ActInputValue item in mAct.UpdateOperationInputValues)
-            {
-                item.Value = item.Param;
-                item.ValueForDriver = VE.Calculate(item.Param);
-            }
+                foreach (ActInputValue item in mAct.UpdateOperationInputValues)
+                {
+                    item.Value = item.Param;
+                    item.ValueForDriver = VE.Calculate(item.Param);
+                }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:10981]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs [hunk 24]
```diff
@@ -605,6 +622,302 @@
...
+        {
+            try
+            {
+               dialogs = e;
+            }
+            catch (Exception ex)
+            {
+                Reporter.ToLog(eLogLevel.ERROR, $"Method - {MethodBase.GetCurrentMethod().Name}, Error - {ex.Message}", ex);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs:822]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs [hunk 25]
```diff
@@ -11100,53 +11092,43 @@
...
-                Reporter.ToLog(eLogLevel.ERROR, "Error in OnNetworkRequestSent ", ex);
+                Reporter.ToLog(eLogLevel.ERROR, $"Method - {MethodBase.GetCurrentMethod().Name}, Error - {ex.Message}", ex);
             }
-        }
-
-        private bool IsToMonitorAllUrls()
-        {
-            return mAct.GetOrCreateInputParam(nameof(ActBrowserElement.eMonitorUrl)).Value == ActBrowserElement.eMonitorUrl.AllUrl.ToString();
-        }
+}
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs [hunk 26]
```diff
@@ -605,6 +621,305 @@ public static bool IsFrameLocatorSupported(eLocateBy locateBy)
...
+                                                                                                                                    ReferenceLoopHandling = ReferenceLoopHandling.Ignore
+                                                                                                                                })
+                ));
+                }
+            }
+            catch (Exception ex)
+            {
+                Reporter.ToLog(eLogLevel.ERROR, $"Method - {MethodBase.GetCurrentMethod().Name}, Error - {ex.Message}", ex);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserTab.cs:761]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs [hunk 27]
```diff
@@ -11100,53 +11093,43 @@ private void OnNetworkRequestSent(object sender, NetworkRequestSentEventArgs e)
         {
             try
             {
-                if (IsToMonitorAllUrls() || IsToMonitorOnlySelectedUrls(e.RequestUrl))
+                if (_BrowserHelper.ShouldMonitorAllUrls() || _BrowserHelper.ShouldMonitorUrl( e.RequestUrl))
                 {
-                    networkRequestLogList.Add(new Tuple<string, object>("RequestUrl:" + e.RequestUrl, JsonConvert.SerializeObject(e)));
+                    networkRequestLogList.Add(new Tuple<string, object>($"RequestUrl: {e.RequestUrl}", JsonConvert.SerializeObject(e, Formatting.Indented)));
                 }
+
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:11100]: no need

**Review by Maheshkale447 — APPROVED** [review]

**Pull request merged**